htmd/parameterization/dihedral.py

```diff
@@ -0,0 +1,548 @@
+# (c) 2015-2017 Acellera Ltd http://www.acellera.com
+# All Rights Reserved
+# Distributed under HTMD Software License Agreement
+# No redistribution in whole or part
+#
+import os
+import logging
+import unittest
+import numpy as np
+import nlopt
+import matplotlib
+matplotlib.use('Agg')
+import matplotlib.pyplot as plt
+from sklearn.linear_model import LinearRegression
+
+from htmd.molecule.util import dihedralAngle
+from htmd.parameterization.ffevaluate import FFEvaluate
+
+logger = logging.getLogger(__name__)
+
+
+class DihedralFitting:
+    """
+    Dihedral parameter fitting
+
+    Capabilities
+    ------------
+    - Fit parameters from QM energies
+    - Fit multiple dihedral angles simulatanously
+    - Global parameter optimizer
+
+    Attributes
+    ----------
+    molecule : FFMolecule
+        Molecule object
+    dihedrals : list of lists
+        List of dihedral angles. Each angle is define by 4 atom indices.
+    qm_results : List of QMResult
+        Reference QM results. The number of results has to be the same to the number of dihedrals.
+    result_directory : str
+        The directory to save plots
+    zerod_parameters : bool
+        If set to True, the initial parameter values are set to zeros, else the existing parameters are used as a guess.
+    """
+
+    MAX_DIHEDRAL_MULTIPLICITY = 6
+
+    def __init__(self):
+
+        self.molecule = None
+        self.dihedrals = []
+        self.qm_results = []
+        self.result_directory = None
+        self.zeroed_paramters = False
+
+        self.parameters = None
+        self.loss = None
+
+        self._names = None
+        self._equivalent_indices = None
+
+        self._valid_qm_results = None
+        self._reference_energies = None
+        self._coords = None
+        self._angle_values = None
+
+        self._initial_energies = None
+        self._target_energies = None
+        self._fitted_energies = None
+
+        self._all_target_energies = None
+        self._angle_values_rad = None
+
+    @property
+    def numDihedrals(self):
+        """Number of dihedral angles"""
+
+        return len(self.dihedrals)
+
+    def _getEquivalentDihedrals(self, dihedral):
+        """
+        Find equivalent dihedral angles to the specificied one.
+        """
+        #TODO maybe this should be moved to FFMolecule
+
+        types = [self.molecule._rtf.type_by_index[index] for index in dihedral]
+
+        all_uses = []
+        for dihedral_indices in self.molecule.dihedrals:
+            dihedral_types = [self.molecule._rtf.type_by_index[index] for index in dihedral_indices]
+            if types == dihedral_types or types == dihedral_types[::-1]:
+                all_uses.append(dihedral_indices)
+
+        # Now for each of the uses, remove any which are equivalent
+        unique_uses = [dihedral]
```

Comment:
`unique_uses` seems like a confusing variable name for a list of equivalent dihedrals

Reply:
Changed!